src/__Private/Visitor/ASTVisitor.hack

```diff
@@ -0,0 +1,278 @@
+namespace Slack\GraphQL\__Private;
+
+use namespace \Graphpinator\Parser;
+
+abstract class ASTVisitor {
+
+    /**
+     * Walk the provided AST, calling the appropriate enter and leave methods as we go.
+     */
+    final public function walk(Parser\ParsedRequest $request): void {
+        $this->visitParsedRequest($request);
+    }
+
+    /**
+     * Run this another visitor, then this visitor, in sequence on each node.
+     */
+    final public function runAfter(ASTVisitor $visitor): ASTVisitor {
+        return new ParallelVisitor(vec[$visitor, $this]);
+    }
+
+    // Actual visitor implemention.
+    //
+    // TODO: Is there a better way to enforce that each node be visitable?
+    // Maybe we should force all Graphphinator Parsed AST nodes to implement
+    // an `accept` method and move this logic there?
+
+    private function visitParsedRequest(Parser\ParsedRequest $node): void {
+        foreach ($node->getOperations() as $operation) {
+            $this->visitOperation($operation);
+        }
+        foreach ($node->getFragments() as $fragment) {
+            // TODO: Fragments
+        }
+    }
+
+    private function visitField(Parser\Field\Field $node): void {
+        $this->enterField($node);
+        $fieldset = $node->getFields();
+        if ($fieldset) {
+            $this->visitFieldSet($fieldset);
+        }
+        foreach ($node->getArguments() ?? dict[] as $argument) {
+            $this->visitArgumentValue($argument);
+        }
+        foreach ($node->getDirectives() ?? vec[] as $directive) {
+            // TODO: Directives
+        }
+        $this->leaveField($node);
+    }
+
+    private function visitFieldSet(Parser\Field\FieldSet $node): void {
+        $this->enterFieldSet($node);
+        foreach ($node->getFields() as $field) {
+            $this->visitField($field);
+        }
+        foreach ($node->getFragmentSpreads() as $fragment_spread) {
+            // TODO: Fragments
+        }
+        $this->leaveFieldSet($node);
+    }
+
+    private function visitOperation(Parser\Operation\Operation $node): void {
+        $this->enterOperation($node);
+        $fieldset = $node->getFields();
+        if ($fieldset) {
+            $this->visitFieldSet($fieldset);
+        }
+        foreach ($node->getDirectives() ?? vec[] as $directive) {
+            // TODO: Directives
+        }
+        foreach ($node->getVariables() ?? vec[] as $variable) {
+            $this->visitVariable($variable);
+        }
+        $this->leaveOperation($node);
+    }
+
+    private function visitTypeRef(Parser\TypeRef\TypeRef $node): void {
```

Comment:
Some of these feel unnecessarily granular to me. Do we need to visit literally every single AST node?

In this particular example, I don't think any validation rule can do anything useful with a TypeRef without any context (e.g. you can't even validate whether a type name is valid without knowing if we're expecting an input type or output type). So a validation rule like "variable types are valid" would likely override `visitVariableDef` and then call `$node->getType()`, instead of overriding `visitTypeRef` directly; or it could even just override `visitOperation` and call `$operation_node->getVariableDefs()`...

Reply:
Yeah good point. Every node is visitable in the reference implementation (I'm pretty sure), so I just mindlessly followed that 😉 . But yeah, I agree that we probably don't need all of these. I'm not sure exactly what we'll need, but I imagine at least the following:

* Operations
* Variables
* Fields
* Fragments
* Directives
* Values (I think?)

But maybe we don't need separate methods for every single type of value, etc.

Reply:
> Every node is visitable in the reference implementation

AFAIK the reference implementation is meant to be more generic, not just support validation -- so it probably allows a lot of cases we don't really need, at least for now.